### PR TITLE
Handled complex types

### PR DESCRIPTION
**Pull request by lsavouillannxw**
The code in this PR handle the complex types found inside the struct we store in the table.

The idea for now is just to store the complex types as json string (running a json.marshal on them) and to unmarshal them at retrieval.
It's the same for slices of complex types, we will store them as a list of string and rebuild the list on retrieval.

### CODE REVIEW

File: type.go [hunk 1]
```diff
@@ -64,6 +115,20 @@ func cassaType(i interface{}) gocql.Type {
 	case Counter:
 		return gocql.TypeCounter
 	}
+	switch reflect.ValueOf(i).Kind() {
```
Comment (fsarron-nexway) [inline type.go]: please a comment to tell why it's necessary to do so

File: type.go [hunk 2]
```diff
@@ -36,11 +38,60 @@ func stringTypeOf(i interface{}) (string, error) {
...
+					if err != nil {
+						return "", fmt.Errorf("Unsupported translation of value type %T", err)
+					}
+					result = append(result, string(txt))
+				}
+				return result, nil
+			}
+			return i, nil
+		case reflect.Map:
+			keyVal := reflect.Indirect(reflect.New(reflect.TypeOf(i).Key())).Interface()
```
Comment (fsarron-nexway) [inline type.go:73]: it seems this switch block is the same as the counterpart in stringTypeOf(), please check

File: type.go [hunk 3]
```diff
@@ -36,11 +38,60 @@ func stringTypeOf(i interface{}) (string, error) {
...
 	}
 	return cassaTypeToString(ct)
 }
 
+type structure struct {
+	A string
+        B int
+}
+
+func ProcessValue(i interface{}) (interface{}, error) {
```
Comment (fsarron-nexway) [inline type.go]: Please tell in comment what is going on in this function

File: type.go [hunk 4]
```diff
@@ -36,11 +38,60 @@ func stringTypeOf(i interface{}) (string, error) {
...
 	ct := cassaType(i)
 	if ct == gocql.TypeCustom {
-		return "", fmt.Errorf("Unsupported type %T", i)
+		ct = gocql.TypeVarchar
+		// return "", fmt.Errorf("Unsupported type %T", i)
 	}
 	return cassaTypeToString(ct)
 }
 
+type structure struct {
```
Comment (fsarron-nexway) [inline type.go]: trash struct ?

File: reflect/reflect.go [hunk 5]
```diff
@@ -41,15 +42,62 @@ func StructToMap(val interface{}) (map[string]interface{}, bool) {
...
+				//fmt.Print("Field set naturally!!!\n")
 				structField.Set(r.ValueOf(v))
+			} else if structField.Kind().String() == r.TypeOf(v).Name() {
+				//fmt.Print("Field set with convert !!!\n")
+				structField.Set(r.ValueOf(v).Convert(structField.Type()))
+			} else {
+				fmt.Print("Please handle these types !!!\n")
 			}
+		} else {
+			//fmt.Printf("field %q not found\n", k) TODO: in which situation do we reach this point? oO
```
Comment (fsarron-nexway) [inline reflect/reflect.go]: you get here when you gave a map m with some key not matching any field in struc.
Please return an error or panic, depending on the gravity of the contract violation ...

File: reflect/reflect.go [hunk 6]
```diff
@@ -41,15 +42,62 @@ func StructToMap(val interface{}) (map[string]interface{}, bool) {
...
 	for k, v := range m {
+		//fmt.Printf("k: %+v  v: %+v\n", k, v)
 		if info, ok := sinfo.FieldsMap[k]; ok {
+			//fmt.Printf("info: %+v\n", info)
 			structField := val.Field(info.Num)
-			if structField.Type().Name() == r.TypeOf(v).Name() {
+			//fmt.Printf("type struct: %q, %q, %q\n", structField.Type(), structField.Type().Name(), structField.Kind())
+			//fmt.Printf("type value: %q\n", r.TypeOf(v).Name())
+			//fmt.Printf("value: %+v\n", r.ValueOf(v))
+			if structField.Kind().String() == "slice" || r.TypeOf(v).Kind().String() == "slice" {
```
Comment (fsarron-nexway) [inline reflect/reflect.go]: This is especially hard to follow
- please set appropriate vars for frequently used method chains (for instance r.TypeOf(v) , r.TypeOf(v).Elem() ...
- must be possible to have nicer names than r,k,v,m,struct,val
- please comment subtleties

**Review by fsarron-nexway — CHANGES_REQUESTED** [review]
Code needs commenting and nicer variable naming to be handled by another dev or the author himself in several weeks

File: type.go [hunk 7]
```diff
@@ -20,7 +21,8 @@ func stringTypeOf(i interface{}) (string, error) {
 			elemVal := reflect.Indirect(reflect.New(reflect.TypeOf(i).Elem())).Interface()
 			ct := cassaType(elemVal)
 			if ct == gocql.TypeCustom {
-				return "", fmt.Errorf("Unsupported type %T", i)
+				ct = gocql.TypeVarchar
+				//return "", fmt.Errorf("Unsupported type %T", i)
```
Comment (spierre) [inline type.go]: remove commented deprecated code and add comment why TypeCustom become TypeVarchar ?

File: reflect/reflect.go (hunk 6 above)
Comment (thureto) [reply reflect/reflect.go]: i'm fine with k, v on my side pretty common to use it but "struc" (c'est quoi struc)

File: reflect/reflect.go [hunk 8]
```diff
@@ -41,15 +42,62 @@ func StructToMap(val interface{}) (map[string]interface{}, bool) {
...
+				structField.Set(r.ValueOf(v).Convert(structField.Type()))
+			} else {
+				fmt.Print("Please handle these types !!!\n")
 			}
+		} else {
+			//fmt.Printf("field %q not found\n", k) TODO: in which situation do we reach this point? oO
 		}
+		//fmt.Printf("Check fill struc: %+v\n", struc)
 	}
 	return nil
```
Comment (thureto) [inline reflect/reflect.go:102]: You don't want to manage errors ? if i'm understand well you return nil all the time

**Review by spierre — APPROVED** [review]

File: reflect/reflect.go (hunk 6 above)
Comment (lsavouillannxw) [reply reflect/reflect.go]: struc interface{} => structure

File: reflect/reflect.go (hunk 5 above)
Comment (lsavouillannxw) [reply reflect/reflect.go]: I can't... for each test we fall in that usecase half of the time :/ but it doesn't seem to be harmful to skip it...

File: reflect/reflect.go (hunk 8 above)
Comment (lsavouillannxw) [reply reflect/reflect.go:102]: fixed

File: type.go (hunk 7 above)
Comment (lsavouillannxw) [reply type.go]: done

File: type.go (hunk 2 above)
Comment (lsavouillannxw) [reply type.go:73]: fixed

File: type.go (hunk 1 above)
Comment (lsavouillannxw) [reply type.go]: done

**Review by fsarron-nexway — APPROVED** [review]
